packages/foundry/contracts/prediction-market/lib/PredictionMarketLib.sol

```diff
@@ -0,0 +1,297 @@
+        // expected to match other pricing methods.
+        return Math.mulDiv(balanceToken0, 1e18, balanceToken1);
+    }
+
+    /**
+     * @notice Whether a market has been initalized
+     * @dev All initalized markets have at minimum an open price
+     * @param self The prediction market
+     */
+    function isInitalized(PredictionMarket memory self) internal pure returns (bool) {
+        return self.openPrice > 0;
+    }
+
+    /**
+     * @notice Whether a market is currently active 
+     * @dev A market is active and can accept swaps / deposits if it is initalized and the
+     * end time has not passed
+     * @param self The prediction market
+     */
+    function isActive(PredictionMarket memory self) internal view returns (bool) {
+        return isInitalized(self) && self.endTime < block.timestamp;
+    }
+
+    /**
+     * @notice Calculate the net available liquidity in the market after fees
+     * @dev Used to calculate quotes since we keep all liquidity in the market until settlement
+     * @param self The prediction market
+     * @return liquidity Calculated net liquidity amount
+     */
+    function netLiquidity(PredictionMarket memory self) internal pure returns (uint256 liquidity) {
+        return self.liquidity - self.fees;
+    }
+
+    /**
+     * @notice Whether swaps can be facilitated in the market
+     * @dev If a market has liquidity on only 1 side, then a swap is not possible
+     * @param self The prediction market
+     */
+    function canSwap(PredictionMarket memory self) internal view returns (bool) {
+        return isActive(self) && self.balanceBear * self.balanceBull != 0;
+    }
+
+    /**
+     * @notice Add liquidity to a given market 
+     * @dev Credits msg.sender with value of bull/bear units and increments liquidity.
+     * @param self The prediction market
+     * @param amount The deposited liquidity amount
+     * @param side The side to add liquidit to
+     * @param feePercentage The fee amount
+     * @return bullAmount The bull amount of units credited
+     * @return bearAmount The bear amount of units credited
+     */
+    function addLiquidity(
+        PredictionMarket memory self,
+        Side side,
+        uint256 amount,
+        uint256 feePercentage
+    ) internal returns (uint256 bullAmount, uint256 bearAmount) {
+        if(side == Side.Both){
+            return _addProportionalLiquidity(self, amount, feePercentage);
+        } else {
+            return _addSingleSidedLiquidity(self, side, amount, feePercentage);
+        }
+    }
+
+    /**
+     * @notice get the fee amount given an amount and fee percentage
+     * @param amount The total amount
+     * @param feePercentage The fee amount
+     * @return fee The calculated fee amount
+     */
+    function _calculateFees(
+        uint256 amount,
+        uint256 feePercentage
+    ) private pure returns (uint256 fee) {
+        return Math.mulDiv(amount, feePercentage, 1e6);
```

Comment:
FixedPoint has a mulDiv too; this is doing some fee scaling factor magic that could stand to be documented.